### PR TITLE
[Fix #5038] Fix false positive for `Performance/RegexpMatch` cop

### LINKED ISSUE #5038: False positive on Performance/RegexpMatch
Performance/RegexpMatch false detects code like this to be a valid candidate for using `.match()` instead of `=~`:

~~~ruby
return Regexp.last_match(1).to_i if uri =~ %r{/m?(\d+)$}
~~~

## Expected behavior

Rubocop should correctly identify the use of `Regexp.last_match`

## Actual behavior

Rubocop suggests to replace `=~` by `.match()`

## Steps to reproduce the problem

Run Rubocop against the code above, with Ruby version 2.4.2:

~~~
➜ ckruse@spinnweb ~/sites/cforum (master) ✗ % cat .ruby-version
2.4.2
➜ ckruse@spinnweb ~/sites/cforum (master) ✗ % cat test.rb
# frozen_string_literal: true

return Regexp.last_match(1).to_i if uri =~ %r{/m?(\d+)$}
➜ ckruse@spinnweb ~/sites/cforum (master) ✗ % rubocop test.rb
Inspecting 1 file
C

Offenses:

test.rb:3:37: C: Use match? instead of =~ when MatchData is not used.
return Regexp.last_match(1).to_i if uri =~ %r{/m?(\d+)$}
                                    ^^^^^^^^^^^^^^^^^^^^

1 file inspected, 1 offense detected
➜ ckruse@spinnweb ~/sites/cforum (master) ✗ %
~~~

## RuboCop version

```
➜ ckruse@spinnweb ~/sites/cforum (master) ✗ % rubocop -V
0.51.0 (using Parser 2.4.0.0, running on ruby 2.4.2 x86_64-darwin17)
➜ ckruse@spinnweb ~/sites/cforum (master) ✗ %
```

### PR DESCRIPTION
**Pull request by koic**
Fixes #5038.

This PR fixes a false positive for `Performance::RegexpMatch` when using `MatchData` before guard clause.

Perhaps other false positives may be considered, but this PR will first solve the problem reported in the above issue.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `rake default` or `rake parallel`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/performance/regexp_match.rb [hunk 1]
```diff
@@ -199,6 +202,10 @@ def scope_root(node)
           end
         end
 
+        def modifier_if_or_unless?(source, condition)
+          source =~ Regexp.new("(if|unless) +#{Regexp.escape(condition)}\\z")
+        end
```
Comment (pocke) [inline lib/rubocop/cop/performance/regexp_match.rb]: Probably you can use `IfNode#modifier_form?`.
https://github.com/bbatsov/rubocop/blob/031b29b495ad435dc69172fb7603042894faae53/lib/rubocop/ast/node/if_node.rb#L76-L83

File: lib/rubocop/cop/performance/regexp_match.rb [hunk 2]
```diff
@@ -154,6 +154,9 @@ def message(node)
         def last_match_used?(match_node)
           scope_root = scope_root(match_node)
           body = scope_root ? scope_body(scope_root) : match_node.ancestors.last
+
+          return true if modifier_if_or_unless?(body.source, match_node.source)
```
Comment (pocke) [inline lib/rubocop/cop/performance/regexp_match.rb]: I guess the implementation has a false positive. The cop will ignore the following code.

```ruby
def foo
  return if /re/ =~ str
  do_something Regexp.last_match[1]
end
```

File: lib/rubocop/cop/performance/regexp_match.rb (hunk 1 above)
Comment (koic) [reply lib/rubocop/cop/performance/regexp_match.rb]: Thanks for the advice. I fixed it.

File: lib/rubocop/cop/performance/regexp_match.rb (hunk 2 above)
Comment (koic) [reply lib/rubocop/cop/performance/regexp_match.rb]: I'm not sure I understand this problem.
Is it this problem even at the time of RuboCop 0.52.1?
I don't understand the expected behavior and the actual behavior 💦

**Comment by bbatsov** [comment]
👍

**Pull request merged**